query-planner-js/src/buildQueryPlan.ts

```diff
@@ -995,6 +996,30 @@ class FetchGroup {
       ...this.otherDependentGroups,
     ];
   }
+
+  mergeDependentGroups(otherDependentGroups: FetchGroup[]) {
+    for (const dependentGroup of otherDependentGroups) {
+      // In order to avoid duplicate fetches, we try to find existing dependent
+      // groups with the same service and merge path first.
+      const existingDependentGroup = this.dependentGroups.find(
+        (group) =>
+          group.serviceName === dependentGroup.serviceName &&
+          deepEqual(group.mergeAt, dependentGroup.mergeAt),
+      );
+      if (existingDependentGroup) {
+        existingDependentGroup.merge(dependentGroup);
+      } else {
+        this.otherDependentGroups.push(dependentGroup);
+      }
+    }
+  }
+
+  merge(otherGroup: FetchGroup) {
+    this.fields.push(...otherGroup.fields);
+    this.requiredFields.push(...otherGroup.requiredFields);
+    this.providedFields.push(...otherGroup.providedFields);
+    this.mergeDependentGroups(otherGroup.dependentGroups);
```

Comment:
I know you noticed this recusing merging solves concrete inefficiencies, but I'm having trouble convincing myself that it's always correct.

Mainly, unless I'm not reading this correctly, this appear to "flatten" the dependent groups, because we merge the dependents of the dependents of the subGroup into the dependents of the parentGroup. And that doesn't sound safe in general.

Say we merge into group A  a dependent group B <- C (group B with a dependent group C). I believe we end up with A <- B and A <- C, meaning C ends up a direct dependent of A. But cannot it be that C has required fields from B? In which case, we'll end up fetching from C with only the data from A (and in parallel of B) and won't get our proper inputs. Is there a hole in that reasoning/a miscomprehension of the code?

Reply:
@pcmanus If we remove the recursion (and ticket it for follow-up work; not block), do we still solve the use-case / reproduction we set out to resolve?  Does it not make things worse for other cases?

Reply:
@martijnwalraven will know better than I do, but I believe we do fix the main cases we set out to resolve without the recursion, but @martijnwalraven apparently noticed some cases where new duplicate fetches were created (which I don't understand so we need to look at those cases more closely).

Reply:
So, it's a tad embarrassing but my comment above is actually incorrect and the recursive call does not "flatten" anything, it does properly recurse in both `this` and `otherGroup` at the same time.

As I mentioned to @martijnwalraven, I think my brain tricked me in reading the line
```graphql
this.mergeDependentGroups(otherGroup.dependentGroups);
```
more as
```graphql
this.merge(otherGroup.dependentGroups);
```
and was like "wait, why are we descending into the dependents of otherGroup but not of this?", but ... bad brain ... bad.

So sorry about that, this actually looks good.

Maybe a minor suggestion would be to make `mergeDependentGroups` take a `FetchGroup` as argument (so it means "merge the dependent groups of `that` group to the dependent groups of `this` group" (minor aside: when dealing with method that act mostly symmetrically on both `this` and a single other arg, I've been a big fan of calling the other argument `that`, both in comments and code; I find it reads very well)). This would make the recursive call be
```graphql
this.mergeDependentGroups(otherGroup);
```
which is more symmetric.

But just a suggestion and not insisting if you prefer it the way it is: it's not because I had a brain fart that others will.